Table of Contents

1. Import Libraries

2. Import Data

3. Using groupby() function find the average numbers of orders by department

4.  Create 'loyalty_flag' column

4.1 Check the basic statistics of the product prices for each loyalty category

5. Create 'spending_flag' column 

6. Create 'frequency_flag' column

7. Export Dataframes

1. Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

2. Import Data

In [2]:
# Create path
path = '/Users/ianfleming/Desktop/05-2025 Instacart Basket Analysis'

In [3]:
# Import merged orders and products dataframe
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'prepared data', 'ords_prods_merge.pkl'))

In [4]:
# Check import
df_ords_prods_merge

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True         196   
1                         8                     NaN         True       14084   
2                         8                     NaN         True       12427   
3                         8                     NaN         True       26088   
4                         8                     NaN         True       26405   
...                     ...                     ...          ...         ...   
32434484                 12                     7.0        False       14197   
32434485                 12                     7.0        False       38730   
32434486                 12                     7.0        False       31477   
32434487                 12                     7.0        False        6567   
32434488                 12                     7.0        False       22920   

          add_to_cart_order  reordered  \
0                         1          0   
1                         2          0   
2                         3          0   
3                         4          0   
4                         5          0   
...                     ...        ...   
32434484                  5          1   
32434485                  6          0   
32434486                  7          0   
32434487                  8          0   
32434488                  9          0   

                                     product_name  aisle_id  department_id  \
0                                            Soda      77.0            7.0   
1         Organic Unsweetened Vanilla Almond Milk      91.0           16.0   
2                             Original Beef Jerky      23.0           19.0   
3                      Aged White Cheddar Popcorn      23.0           19.0   
4                XL Pick-A-Size Paper Towel Rolls      54.0           17.0   
...                                           ...       ...            ...   
32434484                             Tomato Paste       9.0            9.0   
32434485          Brownie Crunch High Protein Bar       3.0           19.0   
32434486    High Protein Bar Chunky Peanut Butter       3.0           19.0   
32434487      Chocolate Peanut Butter Protein Bar       3.0           19.0   
32434488      Roasted & Salted Shelled Pistachios     117.0           19.0   

          prices match    price_range_loc    busiest_days  \
0            9.0  both  Mid–range product  Regularly busy   
1           12.5  both  Mid–range product  Regularly busy   
2            4.4  both  Low-range product  Regularly busy   
3            4.7  both  Low-range product  Regularly busy   
4            1.0  both  Low-range product  Regularly busy   
...          ...   ...                ...             ...   
32434484     5.6  both  Mid–range product    Busiest days   
32434485     5.9  both  Mid–range product    Busiest days   
32434486     4.2  both  Low-range product    Busiest days   
32434487     4.9  both  Low-range product    Busiest days   
32434488     1.7  both  Low-range product    Busiest days   

         busiest_period_of_day  max_order      loyalty_flag  
0               Average orders 

In [5]:
df_ords_prods_merge.shape

(32434489, 20)

STEP 2. 

3. Using groupby() function find the average numbers of orders by department

In [6]:
# Use groupby() function on product_name column
df_ords_prods_merge.groupby('product_name')

In [8]:
# Find the aggregated mean of 'order_number' column grouped by 'department_id' column
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              15.457838
2.0              17.277920
3.0              17.170395
4.0              17.811403
5.0              15.215751
6.0              16.439806
7.0              17.225802
8.0              15.340650
9.0              15.895474
10.0             20.197148
11.0             16.170638
12.0             15.887671
13.0             16.583536
14.0             16.773669
15.0             16.165037
16.0             17.666284
17.0             15.694469
18.0             19.310514
19.0             17.177343
20.0             16.473447
21.0             22.902379

STEP 3.
When comparing the mean 'order_number' for each department between the full dataset and the subset, there are some small but clear differences. For example, 'department_id' 1 has a mean of 15.46 in the full dataset but drops to 14.79 in the subset, suggesting it is ordered slightly earlier in the subset. Similarly, departments 6 and 12 also show a lower mean in the subset. This could suggest that customers in the subset tend to place orders for some departments earlier in their shopping journey. Department 21 remains the latest-ordered department in both datasets, with only a small change in mean order number.

STEP 4. CREATE LOYALTY FLAG - Previously completed in Exercise 4.8

4. Create 'loyalty_flag' column

STEP 5. Check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 

4.1 Check the basic statistics of the product prices for each loyalty category

In [11]:
# Check statistics of loyal customer's spending habits
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773519  1.0  25.0
New customer      7.801203  1.0  25.0
Regular customer  7.798246  1.0  25.0

Loyal customers spend very slightly less per product on average.
New and regular customers have almost identical average prices.
Minimum and maximum prices are the same across all groups. 
There are no major differences in spending habits between customer types based on price alone.

STEP 6. TARGETING DIFFERENT TYPES OF SPENDERS

5. Create 'spending_flag' column

In [13]:
# Create a new column for each user's average price using transform
df_ords_prods_merge['avg_price_per_user'] = df_ords_prods_merge.groupby('user_id')['prices'].transform('mean')

In [14]:
# Use an if statement to assign the spending flag
spending_flag = []

for price in df_ords_prods_merge['avg_price_per_user']:
    if price < 10:
        spending_flag.append("Low spender")
    else:
        spending_flag.append("High spender")

In [15]:
# Create column 'spending_flag'
df_ords_prods_merge['spending_flag'] = spending_flag

In [22]:
# Check frequency
df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     32314630
High spender      119859
Name: count, dtype: int64

In [16]:
df_ords_prods_merge.shape

(32434489, 22)

In [17]:
df_ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
5    2398795        1             2                   3                  7   
6    2398795        1             2                   3                  7   
7    2398795        1             2                   3                  7   
8    2398795        1             2                   3                  7   
9    2398795        1             2                   3                  7   
10   2398795        1             2                   3                  7   
11    473747        1             3                   3                 12   
12    473747        1             3                   3                 12   
13    473747        1             3                   3                 12   
14    473747        1             3                   3                 12   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                      NaN         True       14084                  2   
2                      NaN         True       12427                  3   
3                      NaN         True       26088                  4   
4                      NaN         True       26405                  5   
5                     15.0        False         196                  1   
6                     15.0        False       10258                  2   
7                     15.0        False       12427                  3   
8                     15.0        False       13176                  4   
9                     15.0        False       26088                  5   
10                    15.0        False       13032                  6   
11                    21.0        False         196                  1   
12                    21.0        False       12427                  2   
13                    21.0        False       10258                  3   
14                    21.0        False       25133                  4   

    reordered  ... department_id  prices  match    price_range_loc  \
0           0  ...           7.0     9.0   both  Mid–range product   
1           0  ...          16.0    12.5   both  Mid–range product   
2           0  ...          19.0     4.4   both  Low-range product   
3           0  ...          19.0     4.7   both  Low-range product   
4           0  ...          17.0     1.0   both  Low-range product   
5           1  ...           7.0     9.0   both  Mid–range product   
6           0  ...          19.0     3.0   both  Low-range product   
7           1  ...          19.0     4.4   both  Low-range product   
8           0  ...           4.0    10.3   both  Mid–range product   
9           1  ...          19.0     4.7   both  Low-range product   
10          0  ...          14.0     4.0   both  Low-range product   
11          1  ...           7.0     9.0   both  Mid–range product   
12          1  ...          19.0     4.4   both  Low-range product   
13          1  ...          19.0     3.0   both  Low-range product   
14          0  ...          16.0     8.6   both  Mid–range product   

      busiest_days busiest_period_of_day max_order  loyalty_flag  \
0   Regularly busy        Average orders        10  New customer   
1   Regularly busy        Average orders        10  New customer   
2   Regularly busy        Average orders        10  New customer   
3   Regularly busy        Average orders        10  New customer   
4   Regularly busy        Average orders        10  New customer   
5     Slowest days        Average orders 

In [18]:
df_ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
..       ...      ...           ...                 ...                ...   
95    738281        2             4                   2                 10   
96   1673511        2             5                   3                 11   
97   1673511        2             5                   3                 11   
98   1673511        2             5                   3                 11   
99   1673511        2             5                   3                 11   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                      NaN         True       14084                  2   
2                      NaN         True       12427                  3   
3                      NaN         True       26088                  4   
4                      NaN         True       26405                  5   
..                     ...          ...         ...                ...   
95                     8.0        False       21150                 13   
96                     8.0        False       47144                  1   
97                     8.0        False        5322                  2   
98                     8.0        False       17224                  3   
99                     8.0        False       38656                  4   

    reordered  ... department_id  prices  match    price_range_loc  \
0           0  ...           7.0     9.0   both  Mid–range product   
1           0  ...          16.0    12.5   both  Mid–range product   
2           0  ...          19.0     4.4   both  Low-range product   
3           0  ...          19.0     4.7   both  Low-range product   
4           0  ...          17.0     1.0   both  Low-range product   
..        ...  ...           ...     ...    ...                ...   
95          0  ...           1.0     5.9   both  Mid–range product   
96          0  ...          16.0    14.0   both  Mid–range product   
97          0  ...          19.0     2.9   both  Low-range product   
98          0  ...          19.0     1.6   both  Low-range product   
99          0  ...          19.0     3.4   both  Low-range product   

      busiest_days busiest_period_of_day max_order      loyalty_flag  \
0   Regularly busy        Average orders        10      New customer   
1   Regularly busy        Average orders        10      New customer   
2   Regularly busy        Average orders        10      New customer   
3   Regularly busy        Average orders        10      New customer   
4   Regularly busy        Average orders        10      New customer   
..             ...                   ...       ...               ...   
95  Regularly busy           Most orders        14  Regular customer   
96    Slowest days           Most orders        14  Regular customer   
97    Slowest days           Most orders        14  Regular customer   
98    Slowest days           Most orders        14  Regular customer   
99    Slowest days           Most orders        14  Regular customer   

    avg_price_per_user spending_flag  
0             6.367797   Low spender  
1             6.367797   Low spender  
2             6.367797   Low spender  
3             6.367797   Low spender  
4             6.367797   Low spender  
..                 ...           ...  
95            7.515897   Low spender  
96            7.515897   Low spender  
97            7.515897   Low spender  
98            7.515897   Low spender  
99            7.515897   Low spender  

[10

In [19]:
# Tell Panadas not to assign a max to the number of rows it displays
pd.options.display.max_rows = None

In [ ]:
# Check the end of dataframe too and all 100 rows can be displayed
df_ords_prods_merge.tail(100)

STEP 7. FREQUENT VERSUS NON-FREQUENT CUSTOMERS

In [23]:
# Create a new column with each user's median days since prior order
df_ords_prods_merge['median_days_per_user'] = df_ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform('median')

6. Create 'frequency_flag' column

In [24]:
# Create the frequency flag using a for loop
order_frequency_flag = []

for days in df_ords_prods_merge['median_days_per_user']:
    if days > 20:
        order_frequency_flag.append("Non-frequent customer")
    elif days > 10:
        order_frequency_flag.append("Regular customer")
    else:
        order_frequency_flag.append("Frequent customer")

In [25]:
# Add the new column to the dataframe
df_ords_prods_merge['order_frequency_flag'] = order_frequency_flag

In [26]:
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21577653
Regular customer          7217190
Non-frequent customer     3639646
Name: count, dtype: int64

In [27]:
df_ords_prods_merge.shape

(32434489, 24)

In [29]:
df_ords_prods_merge.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   
5   2398795        1             2                   3                  7   
6   2398795        1             2                   3                  7   
7   2398795        1             2                   3                  7   
8   2398795        1             2                   3                  7   
9   2398795        1             2                   3                  7   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   
5                    15.0        False         196                  1   
6                    15.0        False       10258                  2   
7                    15.0        False       12427                  3   
8                    15.0        False       13176                  4   
9                    15.0        False       26088                  5   

   reordered  ... match    price_range_loc    busiest_days  \
0          0  ...  both  Mid–range product  Regularly busy   
1          0  ...  both  Mid–range product  Regularly busy   
2          0  ...  both  Low-range product  Regularly busy   
3          0  ...  both  Low-range product  Regularly busy   
4          0  ...  both  Low-range product  Regularly busy   
5          1  ...  both  Mid–range product    Slowest days   
6          0  ...  both  Low-range product    Slowest days   
7          1  ...  both  Low-range product    Slowest days   
8          0  ...  both  Mid–range product    Slowest days   
9          1  ...  both  Low-range product    Slowest days   

   busiest_period_of_day max_order  loyalty_flag avg_price_per_user  \
0         Average orders        10  New customer           6.367797   
1         Average orders        10  New customer           6.367797   
2         Average orders        10  New customer           6.367797   
3         Average orders        10  New customer           6.367797   
4         Average orders        10  New customer           6.367797   
5         Average orders        10  New customer           6.367797   
6         Average orders        10  New customer           6.367797   
7         Average orders        10  New customer           6.367797   
8         Average orders        10  New customer           6.367797   
9         Average orders        10  New customer           6.367797   

  spending_flag  median_days_per_user   order_frequency_flag  
0   Low spender                  20.5  Non-frequent customer  
1   Low spender                  20.5  Non-frequent customer  
2   Low spender                  20.5  Non-frequent customer  
3   Low spender                  20.5  Non-frequent customer  
4   Low spender                  20.5  Non-frequent customer  
5   Low spender                  20.5  Non-frequent customer  
6   Low spender                  20.5  Non-frequent customer  
7   Low spender                  20.5  Non-frequent customer  
8   Low spender                  20.5  Non-frequent customer  
9   Low spender                  20.5  Non-frequent customer  

[10 rows x 24 columns]

7. Export Dataframes

In [30]:
# Export dataframe with new column
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))